### PR TITLE
accept FileSource options object

### PR DESCRIPTION
**Pull request by mikemorris**
Instead of requiring inheritence from mbgl.FileSource, refs #141

Would love a review from @tmpsantos here, as this currently breaks both of the [request coalescing](https://github.com/mapbox/node-mapbox-gl-native/blob/396fcc8b8ccd26fc4d2ac3bcc1345548cce47e0c/test/js/map.test.js#L162-L298) tests. However, the current implementation of request coalescing may be responsible for https://github.com/mapbox/tilelive-gl/issues/37. I'm unsure what the proper implementation should look like here.

### CODE REVIEW

**Comment by tmpsantos** [comment]
> as this currently breaks both of the request coalescing tests.

That is correct, I had a look at the patch and it indeed breaks coalescing but not completely.

Request coalescing works withing the same `FileSource` but after this patch we will create on `FileSource` per `NodeMap`. Multiple request to the same URL within the same `NodeMap` will continue to merge into one but I don't think that is going to happen very often (or if it happen at all).

So if we land this patch, we can also remove the tests because they are no longer valid.

**Comment by mikemorris** [comment]
Okay, pinging @jfirebaugh for a strategy review here. Is it worthwhile to lose the request coalescing between maps with a shared `FileSource` for this API change? This patch implements a unique `NodeFileSource` per `NodeMap` (instead of sharing one `FileSource` between many maps) which will increase memory usage too (although file descriptors for libuv `Map` loops is my current system resource bottleneck).

**Comment by jfirebaugh** [comment]
I think we should removing internal request coalescing entirely. If it's critical for a particular node-mapbox-gl-native use case, it can be more easily and safely implemented in the JS-side implementation of the `request` function, and this function can passed to multiple `mbgl.Maps` to obtain cross-map request coalescing.

File: src/node_file_source.cpp [hunk 1]
```diff
@@ -152,14 +127,20 @@ void NodeFileSource::processCancel(const mbgl::Resource& resource) {
             queue->unref();
         }
 
-        auto handle = NanObjectWrapHandle(this);
-        if (handle->Has(NanNew("cancel"))) {
+#if (NODE_MODULE_VERSION > NODE_0_10_MODULE_VERSION)
+        auto optionsObject = v8::Local<v8::Object>::New(v8::Isolate::GetCurrent(), options);
+        if (optionsObject->Has(NanNew("cancel"))) {
```
Comment (jfirebaugh) [inline src/node_file_source.cpp]: `NodeMap::New` validates the presence of a `cancel` function property, so it looks like this check is redundant.

File: src/node_file_source.cpp [hunk 2]
```diff
@@ -119,15 +89,20 @@ void NodeFileSource::processAdd(const mbgl::Resource& resource) {
...
 
-    auto handle = NanObjectWrapHandle(this);
-    auto requestHandle = NanNew<v8::Object>(NodeRequest::Create(handle, resource));
+    auto requestHandle = NanNew<v8::Object>(NodeRequest::Create(this, resource));
 
     v8::Persistent<v8::Object> requestPersistent;
     NanAssignPersistent(requestPersistent, requestHandle);
     pending.emplace(resource, std::move(requestPersistent));
 
+#if (NODE_MODULE_VERSION > NODE_0_10_MODULE_VERSION)
```
Comment (jfirebaugh) [inline src/node_file_source.cpp]: Is there a NAN facility for avoiding this version check?

Comment (mikemorris) [inline src/node_file_source.cpp]: Looks like this was added relatively recently https://github.com/nodejs/nan#persistent-references

**Comment by mikemorris** [comment]
> I think we should removing internal request coalescing entirely. If it's critical for a particular node-mapbox-gl-native use case, it can be more easily and safely implemented in the JS-side implementation of the request function, and this function can passed to multiple mbgl.Maps to obtain cross-map request coalescing.

I like this plan :+1:

File: src/node_file_source.cpp (hunk 1 above)
Comment (mikemorris) [inline src/node_file_source.cpp]: Hrm, should `cancel` actually be optional though?

File: src/node_file_source.cpp (hunk 2 above)
Comment (mikemorris) [inline src/node_file_source.cpp]: Nan::Persistent only just landed in `nan@2.0.0` four days ago. Punting on this for now as it's a pretty significant overhaul.

File: src/node_file_source.cpp (hunk 1 above)
Comment (jfirebaugh) [inline src/node_file_source.cpp]: My gut says yes. If it's not, most implementations will probably be a no-op anyway.

Comment (mikemorris) [inline src/node_file_source.cpp]: Dropped `cancel` requirement from constructor instead.

**Comment by jfirebaugh** [comment]
LGTM

**Pull request merged**